### PR TITLE
feat: implemented tracking as per spec

### LINKED ISSUE #1164: [FEATURE] Implement Tracking in Java SDK
Implement tracking as per spec: https://github.com/open-feature/spec/commit/cd99c3560a68c1810971be19b87ea803fb9b0b3a

Functional requirements: 

- implement all the features described [here](https://github.com/open-feature/spec/commit/cd99c3560a68c1810971be19b87ea803fb9b0b3a)

Non functional requirements:

- use as many existing abstractions and implementations as possible
  - ie: `tracking event details` should likely use similar underlying functionality to evaluation context / structure
- provider interface(s) are modified to _optionally_ support tracking; not all providers will support it
- associated testing and README entry (see markdown snippet below)
- no breaking changes

README addition (be careful to correctly link to #hooks and #providers and update the code snippet for the language in question

```md
### Tracking

The tracking API allows you to use OpenFeature abstractions and objects to associate user actions with feature flag evaluations.
This is essential for robust experimentation powered by feature flags.
For example, a flag enhancing the appearance of a UI component might drive user engagement to a new feature; to test this hypothesis, telemetry collected by a hook(#hooks) or provider(#providers) can be associated with telemetry reported in the client's `track` function.

{example-language-in-question}

Note that some providers may not support tracking; check the documentation for your provider for more information.
```

### PR DESCRIPTION
**Pull request by warber**
## This PR

- adds support for Tracking as per Spec

### Related Issues

Fixes #1164

### CODE REVIEW

File: src/main/java/dev/openfeature/sdk/OpenFeatureClient.java
Comment (aepfli) [inline src/main/java/dev/openfeature/sdk/OpenFeatureClient.java:3]: please no star imports

**Review by aepfli — CHANGES_REQUESTED** [review]
Thank you the changes look good, but please remove the start imports, and (personal preference) use assertJ for assertions. Plus the comment i wrote regarding the assertCode

File: src/main/java/dev/openfeature/sdk/OpenFeatureClient.java
Comment (warber) [reply src/main/java/dev/openfeature/sdk/OpenFeatureClient.java]: Thanks! I think the IDE "optimized" this. I need to figure out how to configure IntelliJ to avoid wildcard imports for regular files while still allowing them for test files. This is probably still useful for things like static imports of Mockito assertions in tests...

File: src/test/java/dev/openfeature/sdk/TrackingSpecTest.java
Comment (warber) [reply src/test/java/dev/openfeature/sdk/TrackingSpecTest.java]: I think the intention was to explicitly "test" that the API aligns with the spec design. I wasn’t aware of AssertJ—thanks!

File: src/main/java/dev/openfeature/sdk/OpenFeatureClient.java
Comment (aepfli) [reply src/main/java/dev/openfeature/sdk/OpenFeatureClient.java]: i started with https://github.com/open-feature/java-sdk/pull/1098 i should finish it, to prevent this in the future

**Review by aepfli — APPROVED** [review]

**Comment by toddbaert** [comment]
Will review tomorrow!

**Comment by lukas-reining** [comment]
I will also have a look tomorrow if I find time :)

File: src/main/java/dev/openfeature/sdk/FeatureProvider.java
Comment (toddbaert) [inline src/main/java/dev/openfeature/sdk/FeatureProvider.java:3]: We don't allow usage of javax. We actually removed it from the SDK some time ago because it's very explicitly not allowed at some orgs, but it comes in transitively to this repo. Believe it or not, the `Nullable` annotation is present at runtime too.

File: src/main/java/dev/openfeature/sdk/MutableTrackingEventDetails.java
Comment (toddbaert) [inline src/main/java/dev/openfeature/sdk/MutableTrackingEventDetails.java:17]: I think to be consistent with the rest if the SDK we should also implement a `ImutableTrackingEventDetails`.

Not exactly a blocker for me but I think it would be a good idea.

File: src/main/java/dev/openfeature/sdk/OpenFeatureClient.java
Comment (toddbaert) [inline src/main/java/dev/openfeature/sdk/OpenFeatureClient.java:14]: Same as https://github.com/open-feature/java-sdk/pull/1228/files#r1860724481

File: src/main/java/dev/openfeature/sdk/FeatureProvider.java
Comment (aepfli) [reply src/main/java/dev/openfeature/sdk/FeatureProvider.java:3]: Out of curiosity are there other classes which should not be used, might be worth adding archUnit tests preventing those imports. But we should define a list for that.

Comment (warber) [reply src/main/java/dev/openfeature/sdk/FeatureProvider.java]: Dropped the @Nullable hint (even though it would've been a nice touch). 
Maybe we could consider creating our own annotation for that?

File: README.md
Comment (justinabrahms) [inline README.md:225]:
```suggestion
api.getClient().track("visited-promo-page", new MutableTrackingEventDetails(99.77).add("currencyCode", "USD"))
```

File: src/main/java/dev/openfeature/sdk/FeatureProvider.java
Comment (toddbaert) [reply src/main/java/dev/openfeature/sdk/FeatureProvider.java]: > Dropped the @nullable hint (even though it would've been a nice touch). Maybe we could consider creating our own annotation for that?

Not a bad idea, I think. If we could keep it to compile time. But I don't think it would have as much utility as the javax ones since code quality tools wouldn't notice it - still might be helpful in IDE's though.

File: src/main/java/dev/openfeature/sdk/OpenFeatureClient.java
Comment (toddbaert) [inline src/main/java/dev/openfeature/sdk/OpenFeatureClient.java:235]: https://openfeature.dev/specification/sections/flag-evaluation#requirement-1410 says:

> Methods, functions, or operations on the client MUST NOT throw exceptions, or otherwise abnormally terminate. Flag evaluation calls must always return the default value in the event of abnormal execution. Exceptions include functions or methods for the purposes for configuration or setup.

However, this might be an exception, because:

- it's not a transient error - it's essentially a validation; if this throws it will probably never stop throwing until you fix your code
- it's not an evaluation, and instead a side-effect, and presumably wouldn't break the app totally if it threw, even if you didn't wrap it in a try/catch

**Review by toddbaert — APPROVED** [review]
Approving - I'd love to hear from @justinabrahms before merging though.

File: src/main/java/dev/openfeature/sdk/OpenFeatureClient.java
Comment (warber) [reply src/main/java/dev/openfeature/sdk/OpenFeatureClient.java:235]: I agree for exactly the reasons you mentioned. And I've updated the README.md to include this.

File: src/test/java/dev/openfeature/sdk/TrackingSpecTest.java
Comment (chrfwow) [inline src/test/java/dev/openfeature/sdk/TrackingSpecTest.java:36]:
```suggestion
    @Specification(number = "6.1.1", text = "The `client` MUST define a function for tracking the occurrence of " +
```
I think we do not need the last digit

File: src/test/java/dev/openfeature/sdk/MutableTrackingEventDetailsTest.java
Comment (chrfwow) [inline src/test/java/dev/openfeature/sdk/MutableTrackingEventDetailsTest.java:17]: Shouldn't the default value be `null` per the spec?

File: src/main/java/dev/openfeature/sdk/ImmutableTrackingEventDetails.java
Comment (lukas-reining) [inline src/main/java/dev/openfeature/sdk/ImmutableTrackingEventDetails.java:23]: I think there `0` is not the right default. 
https://openfeature.dev/specification/sections/tracking#requirement-621 defines this as optional, so I think a numeric 0 is semantically wrong if no value is given. (same counts for `MutableTrackingEventDetails`)

**Review by lukas-reining — COMMENTED** [review]
Looks good to me!
I only have one concerns about the numeric value.

File: src/test/java/dev/openfeature/sdk/MutableTrackingEventDetailsTest.java
Comment (toddbaert) [reply src/test/java/dev/openfeature/sdk/MutableTrackingEventDetailsTest.java:17]: Ya, as @lukas-reining pointed out, we should consider this point carefully: https://github.com/open-feature/java-sdk/pull/1228#discussion_r1870292488

File: src/main/java/dev/openfeature/sdk/ImmutableTrackingEventDetails.java
Comment (warber) [reply src/main/java/dev/openfeature/sdk/ImmutableTrackingEventDetails.java]: yeah good point, I've changed it to `null` and let the getter return an `Optional<Number>`

Comment (toddbaert) [reply src/main/java/dev/openfeature/sdk/ImmutableTrackingEventDetails.java]: That's a good solution. We already use optional and I think this conveys what we want.

**Comment by toddbaert** [comment]
Since this is an SDK change, I'ld like to get a few more approvals before merge.

File: src/main/java/dev/openfeature/sdk/ImmutableTrackingEventDetails.java
Comment (lukas-reining) [reply src/main/java/dev/openfeature/sdk/ImmutableTrackingEventDetails.java]: Nice that is a good solution!

**Review by lukas-reining — APPROVED** [review]
Looks great! Thanks!

**Review by chrfwow — APPROVED** [review]

File: src/test/java/dev/openfeature/sdk/TrackingSpecTest.java
Comment (toddbaert) [reply src/test/java/dev/openfeature/sdk/TrackingSpecTest.java:38]: Technically this is correct: https://openfeature.dev/specification/sections/tracking/#conditional-requirement-6111

**Comment by toddbaert** [comment]
@warber  Thanks so much! Merging - I added one small change because I was unclear before when I said:

> We might want to add to the docs for .track that it can throw this exception

I meant the Javadocs (though it's fine to put it in the README too). I added this just now: https://github.com/open-feature/java-sdk/pull/1228/commits/48187b57c24a57e947388d1cc37feecc0f1b415f

**Review by beeme1mr — APPROVED** [review]

**Pull request merged**
